File: frontend/src/types/field-sales.ts
/**
 * Field Sales types for TypeScript
 */

export enum SalesRepStatus {
  ACTIVE = 'active',
  INACTIVE = 'inactive',
  ON_LEAVE = 'on_leave',
  SUSPENDED = 'suspended',
}

export enum CheckInStatus {
  PENDING = 'pending',
  APPROVED = 'approved',
  REJECTED = 'rejected',
}

export enum CreditStatus {
  PENDING = 'pending',
  APPROVED = 'approved',
  REJECTED = 'rejected',
  EXPIRED = 'expired',
}

export enum BadgeType {
  BRONZE = 'bronze',
  SILVER = 'silver',
  GOLD = 'gold',
  PLATINUM = 'platinum',
}

export interface SalesRep {
  id: string;
  user_id: string;
  tenant_id: string;
  rep_code: string;
  territory?: string;
  status: SalesRepStatus;
  monthly_target: number;
  current_month_sales: number;
  total_sales: number;
  commission_rate: number;
  total_commission: number;
  points: number;
  level: number;
  badges?: string[];
  csat_score?: number;
  total_visits: number;
  successful_visits: number;
  created_at: string;
  updated_at?: string;
  last_active_at?: string;
}

export interface OutletCheckIn {
  id: string;
  sales_rep_id: string;
  outlet_name: string;
  outlet_id?: string;
  latitude: number;
  longitude: number;
  address?: string;
  geofence_verified: boolean;
  distance_from_outlet?: number;
  photo_url?: string;
  check_in_time: string;
  check_out_time?: string;
  duration_minutes?: number;
  status: CheckInStatus;
  visit_notes?: string;
  products_discussed?: string[];
  orders_placed?: string[];
  created_at: string;
}

export interface VoiceOrder {
  id: string;
  sales_rep_id: string;
  language: string;
  audio_duration_seconds?: number;
  raw_transcription?: string;
  processed_text?: string;
  confidence_score?: number;
  extracted_products?: any[];
  extracted_customer?: any;
  order_id?: string;
  processing_status: string;
  error_message?: string;
  created_at: string;
  processed_at?: string;
}

export interface CreditRequest {
  id: string;
  sales_rep_id: string;
  customer_name: string;
  customer_email?: string;
  customer_phone: string;
  customer_business?: string;
  requested_amount: number;
  credit_term_days: number;
  purpose?: string;
  ai_score?: number;
  ai_recommendation?: string;
  risk_factors?: string[];
  status: CreditStatus;
  approved_amount?: number;
  approved_term_days?: number;
  decision_notes?: string;
  decided_at?: string;
  created_at: string;
  updated_at?: string;
  expires_at?: string;
}

export interface LeaderboardEntry {
  rank: number;
  sales_rep_id: string;
  rep_name: string;
  total_sales: number;
  total_orders: number;
  points_earned: number;
  territory?: string;
}

export interface Achievement {
  id: string;
  code: string;
  name: string;
  name_ar: string;
  description?: string;
  description_ar?: string;
  badge_type: BadgeType;
  badge_icon?: string;
  requirement_type: string;
  requirement_value: number;
  points_reward: number;
  is_earned: boolean;
  earned_at?: string;
}

export interface TargetProgress {
  monthly_target: number;
  current_sales: number;
  progress_percentage: number;
  days_remaining: number;
  daily_average_required: number;
}

export interface SalesRepDashboard {
  rep_info: SalesRep;
  target_progress: TargetProgress;
  recent_check_ins: OutletCheckIn[];
  pending_credits: CreditRequest[];
  leaderboard_position: {
    rank: number;
    total_participants: number;
    percentile: number;
  };
  achievements: Achievement[];
  route_summary: {
    total_visits_today: number;
    completed_visits: number;
    in_progress: number;
  };
}

export interface CheckInFormData {
  outlet_name: string;
  outlet_id?: string;
  latitude: number;
  longitude: number;
  address?: string;
  photo_base64?: string;
  visit_notes?: string;
  products_discussed?: string[];
}

export interface VoiceOrderFormData {
  audio_base64: string;
  language: 'ar' | 'en';
  audio_duration_seconds?: number;
}

export interface CreditRequestFormData {
  customer_name: string;
  customer_email?: string;
  customer_phone: string;
  customer_business?: string;
  requested_amount: number;
  credit_term_days: number;
  purpose?: string;
}

export interface OfflineSyncData {
  check_ins?: CheckInFormData[];
  voice_orders?: VoiceOrderFormData[];
  credit_requests?: CreditRequestFormData[];
  last_sync_at?: string;
}

export interface ARProduct {
  id: string;
  name: string;
  name_ar: string;
  description?: string;
  description_ar?: string;
  price: number;
  image_url?: string;
  ar_model_url: string;
  ar_scale: number;
  ar_placement: 'floor' | 'wall';
}

File: backend/app/schemas/field_sales.py
"""
Field Sales API schemas for validation and serialization
"""

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field, validator

from app.models.field_sales import (
    SalesRepStatus,
    CheckInStatus,
    CreditStatus,
    BadgeType,
)


# Enums for API
class SalesRepStatusEnum(str, Enum):
    ACTIVE = "active"
    INACTIVE = "inactive"
    ON_LEAVE = "on_leave"
    SUSPENDED = "suspended"


class CheckInStatusEnum(str, Enum):
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"


class CreditStatusEnum(str, Enum):
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"
    EXPIRED = "expired"


class BadgeTypeEnum(str, Enum):
    BRONZE = "bronze"
    SILVER = "silver"
    GOLD = "gold"
    PLATINUM = "platinum"


# Sales Rep Schemas
class SalesRepBase(BaseModel):
    rep_code: str = Field(..., min_length=3, max_length=20)
    territory: Optional[str] = Field(None, max_length=100)
    monthly_target: Decimal = Field(default=0, ge=0)
    commission_rate: float = Field(default=0.05, ge=0, le=1)


class SalesRepCreate(SalesRepBase):
    user_id: UUID
    manager_id: Optional[UUID] = None


class SalesRepUpdate(BaseModel):
    territory: Optional[str] = None
    monthly_target: Optional[Decimal] = None
    commission_rate: Optional[float] = None
    status: Optional[SalesRepStatusEnum] = None


class SalesRepResponse(SalesRepBase):
    id: UUID
    user_id: UUID
    tenant_id: str
    status: SalesRepStatusEnum
    current_month_sales: Decimal
    total_sales: Decimal
    total_commission: Decimal
    points: int
    level: int
    badges: Optional[List[str]] = None
    csat_score: Optional[float] = None
    total_visits: int
    successful_visits: int
    created_at: datetime
    updated_at: Optional[datetime] = None
    last_active_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Outlet Check-in Schemas
class CheckInLocation(BaseModel):
    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)
    address: Optional[str] = None


class OutletCheckInCreate(BaseModel):
    outlet_name: str = Field(..., min_length=1, max_length=255)
    outlet_id: Optional[str] = Field(None, max_length=100)
    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)
    address: Optional[str] = None
    photo_base64: Optional[str] = None  # Base64 encoded photo
    visit_notes: Optional[str] = None
    products_discussed: Optional[List[str]] = None


class OutletCheckInUpdate(BaseModel):
    check_out_time: Optional[datetime] = None
    visit_notes: Optional[str] = None
    products_discussed: Optional[List[str]] = None
    orders_placed: Optional[List[str]] = None
    status: Optional[CheckInStatusEnum] = None


class OutletCheckInResponse(BaseModel):
    id: UUID
    sales_rep_id: UUID
    outlet_name: str
    outlet_id: Optional[str] = None
    latitude: float
    longitude: float
    address: Optional[str] = None
    geofence_verified: bool
    distance_from_outlet: Optional[float] = None
    photo_url: Optional[str] = None
    check_in_time: datetime
    check_out_time: Optional[datetime] = None
    duration_minutes: Optional[int] = None
    status: CheckInStatusEnum
    visit_notes: Optional[str] = None
    products_discussed: Optional[List[str]] = None
    orders_placed: Optional[List[str]] = None
    created_at: datetime

    class Config:
        from_attributes = True


# Voice Order Schemas
class VoiceOrderCreate(BaseModel):
    audio_base64: str  # Base64 encoded audio
    language: str = Field(default="ar", pattern="^(ar|en)$")
    audio_duration_seconds: Optional[int] = None


class VoiceOrderResponse(BaseModel):
    id: UUID
    sales_rep_id: UUID
    language: str
    audio_duration_seconds: Optional[int] = None
    raw_transcription: Optional[str] = None
    processed_text: Optional[str] = None
    confidence_score: Optional[float] = None
    extracted_products: Optional[List[Dict[str, Any]]] = None
    extracted_customer: Optional[Dict[str, Any]] = None
    order_id: Optional[UUID] = None
    processing_status: str
    error_message: Optional[str] = None
    created_at: datetime
    processed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Credit Request Schemas
class CreditRequestCreate(BaseModel):
    customer_name: str = Field(..., min_length=1, max_length=255)
    customer_email: Optional[str] = Field(None, max_length=255)
    customer_phone: str = Field(..., min_length=10, max_length=20)
    customer_business: Optional[str] = Field(None, max_length=255)
    requested_amount: Decimal = Field(..., gt=0)
    credit_term_days: int = Field(default=30, gt=0, le=365)
    purpose: Optional[str] = None


class CreditRequestUpdate(BaseModel):
    status: Optional[CreditStatusEnum] = None
    approved_amount: Optional[Decimal] = None
    approved_term_days: Optional[int] = None
    decision_notes: Optional[str] = None


class CreditRequestResponse(BaseModel):
    id: UUID
    sales_rep_id: UUID
    customer_name: str
    customer_email: Optional[str] = None
    customer_phone: str
    customer_business: Optional[str] = None
    requested_amount: Decimal
    credit_term_days: int
    purpose: Optional[str] = None
    ai_score: Optional[float] = None
    ai_recommendation: Optional[str] = None
    risk_factors: Optional[List[str]] = None
    status: CreditStatusEnum
    approved_amount: Optional[Decimal] = None
    approved_term_days: Optional[int] = None
    decision_notes: Optional[str] = None
    decided_at: Optional[datetime] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Dashboard Schemas
class SalesRepDashboard(BaseModel):
    rep_info: SalesRepResponse
    target_progress: Dict[str, Any]
    recent_check_ins: List[OutletCheckInResponse]
    pending_credits: List[CreditRequestResponse]
    leaderboard_position: Dict[str, Any]
    achievements: List[Dict[str, Any]]
    route_summary: Dict[str, Any]


class TargetProgress(BaseModel):
    monthly_target: Decimal
    current_sales: Decimal
    progress_percentage: float
    days_remaining: int
    daily_average_required: Decimal


class LeaderboardEntry(BaseModel):
    rank: int
    sales_rep_id: UUID
    rep_name: str
    total_sales: Decimal
    total_orders: int
    points_earned: int
    territory: Optional[str] = None


class LeaderboardResponse(BaseModel):
    period_type: str
    period_start: datetime
    period_end: datetime
    entries: List[LeaderboardEntry]
    my_position: Optional[LeaderboardEntry] = None


class AchievementResponse(BaseModel):
    id: UUID
    code: str
    name: str
    name_ar: str
    description: Optional[str] = None
    description_ar: Optional[str] = None
    badge_type: BadgeTypeEnum
    badge_icon: Optional[str] = None
    requirement_type: str
    requirement_value: int
    points_reward: int
    is_earned: bool = False
    earned_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# Offline Sync Schemas
class OfflineSyncData(BaseModel):
    check_ins: Optional[List[OutletCheckInCreate]] = None
    voice_orders: Optional[List[VoiceOrderCreate]] = None
    credit_requests: Optional[List[CreditRequestCreate]] = None
    last_sync_at: Optional[datetime] = None


class OfflineSyncResponse(BaseModel):
    synced_check_ins: int
    synced_voice_orders: int
    synced_credit_requests: int
    failed_items: List[Dict[str, Any]]
    sync_completed_at: datetime
